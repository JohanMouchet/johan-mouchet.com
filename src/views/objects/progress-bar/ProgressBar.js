// @flow

import * as React from "react";
import cx from "classnames";
import "./ProgressBar.scss";

type Props = {
  label?: Node,
  total?: Node,
  filling: number,
  tooltip?: Node,
  tooltipOnHover?: boolean,
  className?: string | Array | Object
};

const ProgressBar = (props: Props) => {
  const { label, total, filling, tooltip, tooltipOnHover, className } = props;

  const classNames = cx(
    "o-progress-bar",
    className,
    tooltipOnHover && "o-progress-bar--tooltip-hover"
  );

  return (
    <div className={classNames}>
      {label && <span className="o-progress-bar__label">{label}</span>}
      {total && <span className="o-progress-bar__total">{total}</span>}
      <div className="o-progress-bar__container">
        <div
          className="o-progress-bar__filling wow"
          style={{ width: filling + "%" }}
          data-progress-bar-tooltip={tooltip}
        ></div>
      </div>
    </div>
  );
};

ProgressBar.defaultProps = {
  filling: 0
};

export default ProgressBar;
