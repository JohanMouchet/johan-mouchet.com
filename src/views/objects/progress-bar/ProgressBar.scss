@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

.o-progress-bar {
  $root: &;

  margin: t-vertical-rhythm(1) 0;
  outline: 0;

  /* Labels */
  &__labels {
    margin-bottom: t-vertical-rhythm(0.25);
    display: flex;
  }

  &__label {
    margin-right: auto;
    letter-spacing: 0.02em;
  }

  &__details {
    margin-left: t-vertical-rhythm(0.25);
  }

  &__progress,
  &__total {
    color: $s-color-text-dimmed;
  }

  /* Container */
  &__container {
    height: 12px;
    width: 100%;
    padding: 2px;
    border: 2px solid $s-color-gray;
    position: relative;

    @include t-wekkit-hdpi-small-dimension;
  }

  /* Filling */
  &__filling {
    position: relative;
    height: 100%;
    width: 0;
    max-width: 100%;
    background-color: $s-color-secondary;
    border-radius: inherit;
    transition: width 0.5s ease-out;
  }

  /* Details on Hover */
  &--details-on-hover {
    #{$root}__details {
      opacity: 0;
      transition: opacity 0.2s ease-out;
    }

    &:hover,
    &:focus {
      #{$root}__details {
        opacity: 1;
      }
    }
  }
}
