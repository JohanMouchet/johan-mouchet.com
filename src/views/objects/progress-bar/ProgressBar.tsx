import * as React from "react";
import cx from "classnames";
import "./ProgressBar.scss";

type Props = {
  label?: string;
  progress?: string;
  progressPct?: boolean;
  total?: string;
  detailsOnHover?: boolean;
  filling: number;
  className?: string | Array<string> | Object;
};

const ProgressBar = ({
  label,
  progress,
  progressPct,
  total,
  detailsOnHover,
  filling = 0,
  className,
}: Props) => {
  const classNames = cx(
    "o-progress-bar",
    className,
    detailsOnHover && "o-progress-bar--details-on-hover"
  );

  const clampedFilling = Math.min(Math.max(0, parseInt(filling)), 100);

  return (
    <div className={classNames} tabIndex={detailsOnHover && "0"}>
      {(label || progress || total) && (
        <span className="o-progress-bar__labels">
          {label && <span className="o-progress-bar__label">{label}</span>}

          {(progress || total) && (
            <span className="o-progress-bar__details">
              {progress && (
                <span className="o-progress-bar__progress">
                  {progressPct ? `${clampedFilling}%` : progress}
                </span>
              )}

              {progress && total && <>&nbsp;/&nbsp;</>}

              {total && <span className="o-progress-bar__total">{total}</span>}
            </span>
          )}
        </span>
      )}

      <div className="o-progress-bar__container">
        <div
          className="o-progress-bar__filling"
          style={{ width: `${clampedFilling}%` }}
        ></div>
      </div>
    </div>
  );
};

export default ProgressBar;
