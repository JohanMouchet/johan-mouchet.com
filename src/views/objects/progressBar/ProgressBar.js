// @flow

import * as React from "react";
import cx from "./node_modules/classnames";
import "./progress-bar.scss";

type Props = {
  label?: Node,
  total?: Node,
  filling: number,
  tooltip?: Node,
  tooltipOnHover: boolean,
  metadata: {
    contextClass?: string,
    additionalClasses?: Array<string>
  }
};

const ProgressBar = (props: Props) => {
  const { label, total, filling, tooltip, tooltipOnHover, metadata } = props;

  const ProgressBarClass = cx(
    "o-progress-bar",
    metadata.contextClass,
    metadata.additionalClasses,
    {
      "o-progress-bar--tooltip-hover": tooltipOnHover
    }
  );

  return (
    <div className={ProgressBarClass}>
      {label && <span className="o-progress-bar__label">{label}</span>}
      {total && <span className="o-progress-bar__total">{total}</span>}
      <div className="o-progress-bar__container">
        <div
          className="o-progress-bar__filling wow"
          style={`width: ${filling}%`}
          data-progress-bar-tooltip={tooltip}
        ></div>
      </div>
    </div>
  );
};

ProgressBar.defaultProps = {
  filling: 0
};

export { ProgressBar, Props };
