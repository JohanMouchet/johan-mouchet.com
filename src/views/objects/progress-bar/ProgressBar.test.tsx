import React from "react";
import { render } from "@testing-library/react";
import ProgressBar from ".";

it("Renders <ProgressBar /> without crashing", () => {
  expect(
    render(
      <ProgressBar
        label="Label"
        progress="Progress"
        progressPct={false}
        total="Total"
        detailsOnHover
        filling={50}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});
