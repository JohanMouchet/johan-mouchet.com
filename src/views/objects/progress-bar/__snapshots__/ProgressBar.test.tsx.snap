// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <ProgressBar /> without crashing 1`] = `
<div
  className="o-progress-bar o-progress-bar--details-on-hover"
  tabIndex={0}
>
  <span
    className="o-progress-bar__labels"
  >
    <span
      className="o-progress-bar__label"
    >
      Label
    </span>
    <span
      className="o-progress-bar__details"
    >
      <span
        className="o-progress-bar__progress"
      >
        Progress
      </span>
       / 
      <span
        className="o-progress-bar__total"
      >
        Total
      </span>
    </span>
  </span>
  <div
    className="o-progress-bar__container"
  >
    <div
      className="o-progress-bar__filling"
      style={
        Object {
          "width": "50%",
        }
      }
    />
  </div>
</div>
`;
